File: Views/MainWindow.axaml.fs
namespace Starter.Views

open Starter
open Starter.Controls
open Starter.Features

open System.Collections.Generic
open System.Windows.Input
open Vanara.PInvoke

open Avalonia
open Avalonia.Input
open Avalonia.Interactivity
open Avalonia.Controls
open Avalonia.Markup.Xaml
open Avalonia.VisualTree

type MainWindow() as this =
    inherit Window()

    let wndProcCallback =
        Win32Properties.CustomWndProcHookCallback(
            fun (_hWnd: nativeint) (msg: uint32) (_wParam: nativeint) (_lParam: nativeint) _ ->
                if msg = uint User32.WindowMessage.WM_HOTKEY then
                    this.Show()
                0
        )

    do this.InitializeComponent()

    member this.ViewModel =
        match this.DataContext with
        | null -> failwith "No DataContext attached"
        | dc -> dc :?> ViewModels.MainWindowViewModel
    member this.TextBox = this.Get<TextBox> "TextBox"
    member this.SearchEnginePill = this.Get<SearchEnginePill> "SearchEnginePill"

    member this.ResultList = this.Get<ListBox> "ResultList"
    member this.ResultListScrollViewer =
        lazy (this.ResultList.GetVisualDescendants()
              |> Seq.find (fun visual -> visual.Name = "PART_ScrollViewer")
              :?> ScrollViewer)

    member private this.InitializeComponent() =
        AvaloniaXamlLoader.Load(this)
        #if DEBUG
        this.AttachDevTools()
        #endif

        Win32Properties.AddWndProcHookCallback(this, wndProcCallback)

        this.Loaded.Add(fun _ ->
            this.SetupKeyboardShortcuts()

            // Bind background kind
            this.ViewModel.Config
            |> Observable.subscribe (fun config ->
                this.TransparencyLevelHint <-
                    match config.Background with
                    | Config.Background.Acrylic -> [| WindowTransparencyLevel.AcrylicBlur |].AsReadOnly()
                    | Config.Background.Mica -> [| WindowTransparencyLevel.Mica |].AsReadOnly()
                    | Config.Background.None -> [| WindowTransparencyLevel.None |].AsReadOnly()
            )
            |> ignore

            // Bind search results
            this.ResultList.ItemsSource <- this.ViewModel.SearchResults

            // Bind single-search-engine pill
            this.ViewModel.SingleSearchEngineMode
            |> Observable.subscribe (fun singleSeMode ->
                match singleSeMode with
                | None ->
                    this.SearchEnginePill.IsVisible <- false
                    this.SearchEnginePill.DataContext <- null
                | Some se ->
                    this.SearchEnginePill.DataContext <- se
                    this.SearchEnginePill.IsVisible <- true
            )
            |> ignore

            // Subscribe to commands
            this.ViewModel.EmptyTextBoxCommand
            |> Observable.subscribe (fun _ ->
                Threading.Dispatcher.UIThread.Post(fun _ -> this.TextBox.Clear())
            )
            |> ignore
        )

        this.Activated.Add (fun _ ->
            this.TextBox.Focus() |> ignore
            this.TextBox.SelectAll()
            this.ResultList.Selection.Select 0 // Reset selection
            this.ResultListScrollViewer.Value.ScrollToHome() // Scroll to top to preserve the top padding
        )
        #if !DEBUG
        this.Deactivated.Add (fun _ -> this.Hide())
        #endif

    member private this.SetupKeyboardShortcuts() =
        // Subscribe to hide command
        this.ViewModel.HideCommand
        |> Observable.subscribe (fun _ ->
            Threading.Dispatcher.UIThread.Post(fun _ ->
                this.Hide()
            )
        )
        |> ignore

        // Add escape key binding
        let onEscape = // Run when "Escape" is pressed
            ReactiveUI.ReactiveCommand.Create(fun () ->
                if this.ViewModel.SingleSearchEngineMode.Value.IsSome then
                    this.ViewModel.ResetSingleSearchEngineMode()
                else
                    (this.ViewModel.HideCommand :> ICommand).Execute()
            )

        KeyBinding(
            Command = onEscape,
            Gesture = KeyGesture.Parse "Escape"
        )
        |> this.KeyBindings.Add

        // Set keyboard navigation
        // -> The goal is to be able to navigate in the listbox without losing the focus on the textbox
        let r = this.ResultList
        let d = System.EventHandler<KeyEventArgs>(fun _ e ->
            match e.PhysicalKey with
            | PhysicalKey.Backspace ->
                if this.ViewModel.SingleSearchEngineMode.Value.IsSome && this.TextBox.Text = "" then
                    this.ViewModel.ResetSingleSearchEngineMode()
                    e.Handled <- true

            | PhysicalKey.Tab -> // Prevent changing focus
                e.Handled <- true

            | _ ->
                let newSelectedIdx =
                    match e.Key.ToNavigationDirection() |> Option.ofNullable with
                    | Some NavigationDirection.Up ->
                        (r.SelectedIndex - 1)
                        |> max 0
                        |> Some
                    | Some NavigationDirection.Down ->
                        (r.SelectedIndex + 1)
                        |> min (r.ItemCount - 1)
                        |> Some
                    | _ -> None

                match newSelectedIdx with
                | None -> ()
                | Some newSelectedIdx ->
                    if newSelectedIdx = 0 then
                        // Scroll to top to preserve the top padding
                        this.ResultListScrollViewer.Value.ScrollToHome()
                    elif newSelectedIdx = r.ItemCount - 1 then
                        // Scroll to bottom to preserve the bottom padding of the listbox
                        this.ResultListScrollViewer.Value.ScrollToEnd()

                    r.Selection.SelectedIndex <- newSelectedIdx
                    e.Handled <- true
        )
        this.TextBox.AddHandler(InputElement.KeyDownEvent, d, RoutingStrategies.Tunnel)

File: src/Starter.Desktop/Views/MainWindow.axaml.cs
using Avalonia;
using Avalonia.Controls;
using Avalonia.Input;
using Avalonia.Interactivity;
using Avalonia.Threading;
using Avalonia.VisualTree;
using Serilog;
using Starter.Desktop.Controls;
using Starter.Desktop.ViewModels;
using Starter.Features;
using Starter.Features.PlatformInterop;
using R3;

namespace Starter.Desktop.Views;

public partial class MainWindow : TranslucentWindow
{
    private MainWindowViewModel vm = null!;
    private readonly PlatformInterop platformInterop = PlatformInteropFactory.GetPlatformInterop();

    public MainWindow()
    {
        InitializeComponent();
        platformInterop.SetupHotkeyCallback(this);
        TextBox.AddHandler(KeyDownEvent, TextBox_OnKeyDown, RoutingStrategies.Tunnel);

        Activated += (_, _) => OnActivated();
#if !DEBUG
        Deactivated += (_, _) => Hide();
#endif
    }

    protected override void OnLoaded(RoutedEventArgs e)
    {
        base.OnLoaded(e);
        if (DataContext is not MainWindowViewModel dataContext) return;
        vm = dataContext;

        SetupKeyboardShortcuts();

        // Subscribe to view model commands
        vm.ClearTextBox += (_, prefixLength) => Dispatcher.UIThread.Post(() =>
        {
            TextBox.Text = TextBox.Text?[prefixLength..];
            TextBox.CaretIndex -= prefixLength;
        });
        vm.HideWindow += (_, _) => Dispatcher.UIThread.Post(Hide);

        // Bind zoomed mode
        vm.Config
            .Select(config => config.ZoomedMode)
            .DistinctUntilChanged()
            .Subscribe(SetResourceDictionary);

        // Refresh keyboard shortcut when needed
        vm.Config
            .Select(config => config.KeyboardShortcut)
            .DistinctUntilChanged()
            .Subscribe(shortcut => platformInterop.RegisterHotkey(shortcut, this));

        SetResourceDictionary(vm.Config.Value.ZoomedMode);
        platformInterop.RegisterHotkey(vm.Config.Value.KeyboardShortcut, this);
    }

    private void OnActivated()
    {
        // Center window
        var screen = Screens.ScreenFromTopLevel(this) ?? Screens.Primary;
        if (screen is null)
        {
            Log.Error("No screen available. Can't center window");
            return;
        }

        Position =
            screen.WorkingArea.TopLeft
            + new PixelPoint(
                (int)Math.Round((screen.WorkingArea.Width - Width * screen.Scaling) / 2.0),
                (int)Math.Round(screen.WorkingArea.Height * (5.0 / 16.0))
            );

        // Reset focus
        TextBox.Focus();
        TextBox.SelectAll();

        // Reset selection
        ResultList.Selection.Select(0);

        // Scroll to top to preserve the top padding
        if (ResultList.ItemCount != 0)
            ResultList.Scroll?.Offset = new Vector(0, 0); // Scroll to home
    }

    private void SetResourceDictionary(bool zoomedMode)
    {
        var found = this.TryFindResource(
            zoomedMode ? "ZoomedMode" : "NormalMode",
            ActualThemeVariant,
            out var resourceDict
        );

        if (!found || resourceDict is not IResourceDictionary v) return;
        Resources = v;
    }

    private void SetupKeyboardShortcuts()
    {
        // Add escape key binding
        var onEscape = new RelayCommand(() =>
        {
            if (vm.Activator is not null) vm.ResetActivatorCommand.Execute(null);
            else Dispatcher.UIThread.Post(Hide);
        });

        KeyBindings.Add(new KeyBinding()
        {
            Command = onEscape,
            Gesture = new KeyGesture(Key.Escape)
        });
    }

    private void ResultList_OnPointerReleased(object? sender, PointerReleasedEventArgs e)
    {
        var pos = e.GetPosition(this);
        var clickedControl =
            this.GetVisualsAt(pos)
                .FirstOrDefault(v => v.DataContext is SearchResultData);

        vm.SelectResultCommand.Execute(clickedControl?.DataContext);
    }

    private void TextBox_OnKeyDown(object? sender, KeyEventArgs e)
    {
        // Remove activator if caret is at start
        if (e.Key == Key.Back && vm.Activator is not null && TextBox.CaretIndex == 0)
        {
            vm.ResetActivatorCommand.Execute(null);
            e.Handled = true;
            return;
        }

        // Set custom keyboard navigation
        // -> The goal is to be able to navigate in the listbox without losing the focus on the textbox
        int newSelectedIdx;

        if (e.Key == Key.Tab && vm.SearchResults.Count > 0)
        {
            newSelectedIdx = e.KeyModifiers.HasFlag(KeyModifiers.Shift)
                ? Math.Max(ResultList.SelectedIndex - 1, 0)
                : Math.Min(ResultList.SelectedIndex + 1, vm.SearchResults.Count - 1);
        }
        else if (e.Key.ToNavigationDirection() == NavigationDirection.Up)
            newSelectedIdx = Math.Max(ResultList.SelectedIndex - 1, 0);
        else if (e.Key.ToNavigationDirection() == NavigationDirection.Down)
            newSelectedIdx = Math.Min(ResultList.SelectedIndex + 1, vm.SearchResults.Count - 1);
        else return;

        // TODO: Always keep bottom padding
        // Scroll to top or bottom to preserve the paddings
        if (ResultList.Scroll is not null)
        {
            if (newSelectedIdx == 0) ResultList.Scroll.Offset = new Vector(0, 0);
            else if (newSelectedIdx == ResultList.ItemCount - 1) ResultList.Scroll.Offset = new Vector(0, ResultList.Scroll.Extent.Height);
        }

        // Select next item
        ResultList.Selection.SelectedIndex = newSelectedIdx;
        e.Handled = true;
    }
}
